### PR TITLE
Allow components to be executed with a dispatcher

### PR DESCRIPTION
**Pull request by Joe-Palmer**
com_banners converted to use a dispatcher as an example

### CODE REVIEW

**Review by zero-24 — CHANGES_REQUESTED** [review]
Please add the doc blocks to the new com banners dispatcher file.

**Review by zero-24 — CHANGES_REQUESTED** [review]
And please take a look into the travis errors: https://travis-ci.org/joomla/joomla-cms/jobs/194233052

**Comment by SoftForge** [comment]
OK, thanks. I'll add the required docs.

**Comment by zero-24** [comment]
:+1:

File: libraries/cms/component/helper.php [hunk 1]
```diff
@@ -405,6 +429,24 @@ protected static function executeComponent($path)
 	}
 
 	/**
+	 * Dispatch the component.
+	 *
+	 * @param   object  $dispatcher  The dispatcher class.
```
Comment (mbabker) [inline libraries/cms/component/helper.php]: Please typehint the parameter and change the `object` type to `JDispatcherInterface`

File: libraries/cms/dispatcher/dispatcher.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+abstract class JDispatcher implements JDispatcherInterface
+{
+	/**
+	 * The publicly available JApplication
+	 *
+	 * @var     JApplicationCms
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public $app;
```
Comment (mbabker) [inline libraries/cms/dispatcher/dispatcher.php]: Public why?

File: libraries/cms/dispatcher/dispatcher.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	 *
+	 * @param   JApplicationCms  $app  The JApplication for the dispatcher
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function __construct(JApplicationCms $app)
+	{
+		$this->app = $app;
+
+		// Load common and local language files.
```
Comment (mbabker) [inline libraries/cms/dispatcher/dispatcher.php]: Can this logic be moved into the dispatch method?  I don't see a reason language files should be loaded every time a dispatcher is instantiated.

File: libraries/cms/component/helper.php (hunk 1 above)
Comment (SoftForge) [inline libraries/cms/component/helper.php]: OK, done

File: libraries/cms/dispatcher/dispatcher.php (hunk 2 above)
Comment (SoftForge) [inline libraries/cms/dispatcher/dispatcher.php]: I've made it private

File: libraries/cms/dispatcher/dispatcher.php (hunk 3 above)
Comment (SoftForge) [inline libraries/cms/dispatcher/dispatcher.php]: Sure, it's been moved

**Review by mbabker — APPROVED** [review]

File: administrator/components/com_banners/dispatcher.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	/**
+	 * Dispatch method for com_banners
+	 *
+	 * @return  void
+	 *
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public function dispatch()
+	{
+		JHtml::_('behavior.tabstate');
```
Comment (dgt41) [inline administrator/components/com_banners/dispatcher.php]: @SoftForge We shouldn't load assets here, we should always land them in the layout file so users/devs can remove any assets that don't want to load. Actually all these instances are already fixed in the https://github.com/joomla/40-backend-template repo

Comment (laoneo) [inline administrator/components/com_banners/dispatcher.php]: That shouldn't be part of this PR as this code can be found all over Joomla. A new one should be created which moves them in every component.

Comment (SoftForge) [inline administrator/components/com_banners/dispatcher.php]: OK, I'll do another pull request to move tab state behaviour into the relevant views for all the components.

Comment (SoftForge) [inline administrator/components/com_banners/dispatcher.php]: I've removed JHtml::_('behavior.tabstate'); as it will be added in the views with the new backend template when it gets merged tomorrow.
